src/token/src/ledger/history.rs

```diff
@@ -0,0 +1,329 @@
+use crate::types::TxRecord;
+use candid::types::{Compound, Serializer, Type};
+use candid::Principal;
+use ic_cdk::export::candid::{CandidType, Deserialize, Nat};
+use ic_certified_map::{HashTree, RbTree};
+use serde::de::{SeqAccess, Visitor};
+use serde::Deserializer;
+use std::fmt;
+use std::fmt::Formatter;
+
+#[cfg(target_arch = "wasm32")]
+use ic_certified_map::AsHashTree;
+
+const MAX_TREE_SIZE: usize = 100_000;
+
+#[derive(Default, Clone)]
+pub struct History {
+    tree: RbTree<Vec<u8>, Vec<u8>>,
+    tree_size: usize,
+}
+
+impl History {
+    pub fn insert(&mut self, tx: TxRecord) {
+        self.tree
+            .insert(get_key_bytes(&tx.index), get_tx_bytes(&tx));
+        self.tree_size += 1;
+        if self.tree_size > MAX_TREE_SIZE {
+            self.remove_oldest_tx();
+        }
+    }
+
+    fn remove_oldest_tx(&mut self) {
+        let key = self.tree.iter().map(|(key, _)| key).next().unwrap().clone();
+        self.tree.delete(&key);
+        self.tree_size -= 1;
+    }
+
+    pub fn get(&self, id: &Nat) -> Option<TxRecord> {
+        self.tree.get(&get_key_bytes(id)).map(|v| tx_from_bytes(v))
+    }
+
+    pub fn get_range(&self, start: &Nat, limit: &Nat) -> Vec<TxRecord> {
+        fn extract_values(hash_tree: &HashTree, aggr: &mut Vec<TxRecord>) {
+            match hash_tree {
+                HashTree::Fork(boxed) => {
+                    extract_values(&boxed.as_ref().0, aggr);
+                    extract_values(&boxed.as_ref().1, aggr);
+                }
+                HashTree::Leaf(v) => {
+                    aggr.push(tx_from_bytes(&v.to_vec()));
+                }
+                HashTree::Labeled(_, child) => extract_values(child, aggr),
+                _ => {}
+            }
+        }
+
+        let witness = self.tree.value_range(
+            &get_key_bytes(&start.clone()),
+            &get_key_bytes(&(start.clone() + limit.clone() - 1)),
+        );
+        let mut result = vec![];
+        extract_values(&witness, &mut result);
+
+        result
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = TxRecord> + '_ {
+        self.tree.iter().map(|(_, value)| tx_from_bytes(value))
+    }
+
+    pub fn sign(&self) {
+        #[cfg(target_arch = "wasm32")]
+        ic_cdk::api::set_certified_data(&self.tree.root_hash());
+    }
+
+    pub fn get_witness(&self, id: &Nat) -> Option<HashTree> {
+        if self.tree.get(&get_key_bytes(id)).is_some() {
+            Some(self.tree.witness(&get_key_bytes(id)))
+        } else {
+            None
+        }
+    }
```

Comment:
No need to get the key bytes twice as it's the same bytes (saving that one allocation).
The `is_some().then` could be swapped for a `map` instead, which is shorter. Technically it's not mapping the value, but there is no complexity to speak of so either one is fine.
```suggestion
    pub fn get_witness(&self, id: &Nat) -> Option<HashTree> {
        let bytes = get_key_bytes(id);
        self.tree.get(&bytes).is_some().then(|| self.tree.witness(&bytes))
    }
```